feat: allow tag sorting by semver

ldelossa opened the pull request:
Relates to #123.

While this does not introduce "per-branch" tag parsing it does allow an
alternative tag sorting method which maybe a better solution.

With this commit the user can decide to sort the tags by semver instead
of dates.

This is useful where repositories are utilizing a  stable branch model
and back-ports are interleaved with new releases.

For example, if your mainline is on v3.0.0 with it's last release
1/1/2021 and a back-port release of v2.0.1 is released on 1/2/2021,
sorting by semver will correctly order the change log producing
v2.0.0 -> v2.0.1 -> v3.0.0

This functionality is completely opt-in and defaults to the original
"date" sorting

Signed-off-by: ldelossa <louis.delos@gmail.com>

<!-- Thank you for your contribution to git-chglog! Please replace {Please write here} with your description -->


## What does this do / why do we need it?

Explained in commit message.


## How this PR fixes the problem?

Explained in commit message


## What should your reviewer look out for in this PR?

Ensure that there are no logical issues with sorting by semver in respect to the larger git-chlog application. 


## Check lists

* [x] Test passed
* [x] Coding style (indentation, etc)


## Additional Comments (if any)

{Please write here}


## Which issue(s) does this PR fix?

<!--
fixes #
fixes #
-->

ldelossa commented:
@clok can I get a review here?

clok commented on cmd/git-chglog/config.go:
```diff
@@ -138,6 +140,19 @@ func (config *Config) normalizeStyle() {
+	switch {
+	case config.Options.TagSortBy == "":
+		config.Options.TagSortBy = "date"
+	case strings.EqualFold(config.Options.TagSortBy, "date"):
```
👍  for using `strings.EqualFold`

clok commented on tag_reader.go:
```diff
@@ -71,12 +74,37 @@ func (r *tagReader) ReadAll() ([]*Tag, error) {
 	return tags, nil
 }
 
+func filterSemVerTags(tags *[]*Tag) {
+	// filter out any non-semver tags
+	for i, t := range *tags {
+		// remove leading v, since its so
+		// common.
+		name := t.Name
+		if strings.HasPrefix(name, "v") {
+			name = strings.TrimPrefix(name, "v")
+		}
+
+		// attempt semver parse, if not successful
+		// remove it from tags slice.
+		if _, err := semver.NewVersion(name); err != nil {
+			*tags = append((*tags)[:i], (*tags)[i+1:]...)
+		}
+	}
+}
```
For consistnecy, lets use the receiver type of `(*tagReader)` to make this a struct method. There are other methods that modify the `[]*Tag` slice in place for this struct.

```suggestion
func (*tagReader) filterSemVerTags(tags []*Tag) {
	// filter out any non-semver tags
	for i, t := range tags {
		// remove leading v, since its so
		// common.
		name := t.Name
		if strings.HasPrefix(name, "v") {
			name = strings.TrimPrefix(name, "v")
		}

		// attempt semver parse, if not successful
		// remove it from tags slice.
		if _, err := semver.NewVersion(name); err != nil {
			tags = append((tags)[:i], (tags)[i+1:]...)
		}
	}
}
```

ldelossa replied:
Won't this not work? In this example, we are modifying a copy of the tags array, we aren't modifying the original one.

if I run tests with this method, we never replace the original slice header, and the non-semver is included, passed to the semver parse method, and we get a panic. 
```
❯ go test ./...
--- FAIL: TestTagReader (0.00s)
panic: hoge_fuga is not in dotted-tri format [recovered]
	panic: hoge_fuga is not in dotted-tri format

goroutine 160 [running]:
testing.tRunner.func1.2(0x7cf000, 0xc0000915a0)
	/usr/local/go/src/testing/testing.go:1144 +0x332
testing.tRunner.func1(0xc000582300)
	/usr/local/go/src/testing/testing.go:1147 +0x4b6
panic(0x7cf000, 0xc0000915a0)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/coreos/go-semver/semver.Must(...)
	/home/louis/git/gopath/pkg/mod/github.com/coreos/go-semver@v0.3.0/semver/semver.go:65
github.com/coreos/go-semver/semver.New(0xc00048abc0, 0x9, 0x4)
	/home/louis/git/gopath/pkg/mod/github.com/coreos/go-semver@v0.3.0/semver/semver.go:49 +0x77
github.com/git-chglog/git-chglog.(*tagReader).sortTagsBySemver.func1(0x5, 0x4, 0x1)
	/home/louis/git/go/git-chglog/tag_reader.go:159 +0x119
sort.insertionSort_func(0xc000125aa8, 0xc0004967a0, 0x0, 0x6)
	/usr/local/go/src/sort/zfuncversion.go:12 +0xab
sort.quickSort_func(0xc000125aa8, 0xc0004967a0, 0x0, 0x6, 0x6)
	/usr/local/go/src/sort/zfuncversion.go:158 +0x1f6
sort.Slice(0x7b5220, 0xc00033e3f0, 0xc000125aa8)
	/usr/local/go/src/sort/slice.go:20 +0xe8
github.com/git-chglog/git-chglog.(*tagReader).sortTagsBySemver(0xc000125f00, 0xc0002e2b00, 0x6, 0x8)
	/home/louis/git/go/git-chglog/tag_reader.go:156 +0xa5
github.com/git-chglog/git-chglog.(*tagReader).ReadAll(0xc000125f00, 0x0, 0xc000329220, 0x7b5220, 0xc00033e3d8, 0x0)
	/home/louis/git/go/git-chglog/tag_reader.go:83 +0x4a5
github.com/git-chglog/git-chglog.TestTagReader(0xc000582300)
	/home/louis/git/go/git-chglog/tag_reader_test.go:123 +0x117f
testing.tRunner(0xc000582300, 0x874da8)
	/usr/local/go/src/testing/testing.go:1194 +0xef
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1239 +0x2b3
FAIL	github.com/git-chglog/git-chglog	1.104s
ok  	github.com/git-chglog/git-chglog/cmd/git-chglog	(cached)
FAIL
```

clok replied:
Was the call to `r.filterSemVerTags` changed from `filterSemVerTags(&tags)` to `r.filterSemVerTags(tags)` when this test was run?

I think that may be the issue.

clok replied:
@ldelossa my bad. I was confusing the way that the other methods employ the `sort.Slice` method. You are correct that you would need to pass a pointer in order to filter the slice in place.

ldelossa replied:
Np, honestly im a little confused why thats not necessary for sort.Slice? Maybe because it *only* changes the underlying array and not the slice header? 

But regardless, ill def update to taking a receiver.

clok replied:
Yeah, it's something I want to read up on. It feels odd. 

Thanks for following up on this!

clok commented on tag_reader.go:
```diff
@@ -71,12 +74,37 @@ func (r *tagReader) ReadAll() ([]*Tag, error) {
+		r.sortTags(tags)
+		r.assignPreviousAndNextTag(tags)
+	case "semver":
+		filterSemVerTags(&tags)
```
See comment on line 90

```suggestion
		r.filterSemVerTags(&tags)
```

clok reviewed (changes requested):
@ldelossa Thank you for the contribution! I really like the feature. It relates closely with #112 which is for a CLI flag, but I think that the configuration option is a good call as well.

I left one change request regarding the `[]*Tag` slice and method structure. 

I would also like @khos2ow to take a look at this since we have discussed the sorting feature in the past.

clok commented:
For reference, linking to #78 #112

khos2ow commented:
Thank you @ldelossa, this is a good changeset, but I'm having a bit of difficult time understanding why this is needed when #78 seems to be doing the same thing. Can you elaborate a bit of the difference, or why this is needed? Was it because the other pr was stalled?

khos2ow commented:
@clok also with this PR, I really think we have to go through with #112 to better support potentially this PR or any other. For example I think we need to have:

```yaml
sort: "date"
```

instead of

```yaml
options:
  tag_sort_by: "date"
```

ldelossa commented:
> Thank you @ldelossa, this is a good changeset, but I'm having a bit of difficult time understanding why this is needed when #78 seems to be doing the same thing. Can you elaborate a bit of the difference, or why this is needed? Was it because the other pr was stalled?

Yeah, our repo claircore depends on this functionality working, and our change log was broken. I had to fix out regardless so just punted my forks changes here incase you all wanted to merge. 

Was in a rush to fix, but my problem is fixed for now on my own fork. 

If you want me to modify this pr I'm willing to. You can also close it for the ones in flight, i didn't see those before fixing it on a form. 

Just lmk

clok commented:
> > Thank you @ldelossa, this is a good changeset, but I'm having a bit of difficult time understanding why this is needed when #78 seems to be doing the same thing. Can you elaborate a bit of the difference, or why this is needed? Was it because the other pr was stalled?
> 
> Yeah, our repo claircore depends on this functionality working, and our change log was broken. I had to fix it regardless so just punted my forks changes here incase you all wanted to merge.
> 
> Was in a rush to fix, but my problem is fixed for now on my own fork.
> 
> If you want me to modify this pr I'm willing to. You can also close it for the ones in flight, i didn't see those before fixing it on a fork.
> 
> Just lmk.

@ldelossa would you be willing to make the requested changes in my previous review as well as changing the config parameter from `tag_sort_by` to `sort` as @khos2ow suggested? I believe this would put our tool in a good position to merge your contribution and we can pick up #112 pronto.

khos2ow commented:
I think we either should choose this over #78 or the other way around (or both if they are actually complementing each other.) To be honest, I have only skimmed both and haven't dug deep into them, which I will tomorrow.

But at least, it seems @ldelossa is not completely blocked by this and their custom fork is being used. So at least that'll give us a bit of time to fully assess them and plan accordingly.

ldelossa commented:
> > > Thank you @ldelossa, this is a good changeset, but I'm having a bit of difficult time understanding why this is needed when #78 seems to be doing the same thing. Can you elaborate a bit of the difference, or why this is needed? Was it because the other pr was stalled?
> > 
> > 
> > Yeah, our repo claircore depends on this functionality working, and our change log was broken. I had to fix it regardless so just punted my forks changes here incase you all wanted to merge.
> > Was in a rush to fix, but my problem is fixed for now on my own fork.
> > If you want me to modify this pr I'm willing to. You can also close it for the ones in flight, i didn't see those before fixing it on a fork.
> > Just lmk.
> 
> @ldelossa would you be willing to make the requested changes in my previous review as well as changing the config parameter from `tag_sort_by` to `sort` as @khos2ow suggested? I believe this would put our tool in a good position to merge your contribution and we can pick up #112 pronto.

Yeah I'm down. Haven't looked at the other PR but will tomorrow and see what changes you have in mind. 

All other comments in this PR look straight forward enough and LGTM.

ldelossa commented:
@clok @khos2ow I'm down to update where in the config this "sort" tag lives. 

I think you both noted you want the key to be called "sorted". Do you want to keep "date" and "semver" as the possible values?

Also at which level of the config do you want this "sorted" key to appear?

clok commented:
> @clok @khos2ow I'm down to update where in the config this "sort" tag lives. 

Awesome. Thanks @ldelossa !

> I think you both noted you want the key to be called "sort". Do you want to keep "date" and "semver" as the possible values?

Yes, I would like to keep those two options as you have them. 

> Also at which level of the config do you want this "sort" key to appear? 

I feel like this makes sense to be under the general `options` block. 

@khos2ow any thoughts?

ldelossa commented:
I moved all references of "tag_sort_by" in favor of the simpler key "sort".

Moved the tag filtering functions for semver to having a method receiver. 

I'll await further feedback, currently the "sort" key is nested in the highest level under the "options" stanza, which IMO is where I'd put it. 

Let me know any other feedback you all are thinking, I'm not 100% sure if you CLI flag is applicable, since a lot of items in "options" stanza are kept out of the CLI. But just lmk any further changes.

clok reviewed (approved):
I think this is a great addition. Thank you @ldelossa for putting this in place. We can follow up with the CLI flag as we need to decide on how to best resolve the rules for when a config does not align with a flag.

ldelossa commented:
Thanks a lot. Ill watch for the next release which includes this so i can swap our repo over.

khos2ow commented on go.mod:
```diff
@@ -5,6 +5,7 @@ go 1.16
 require (
 	github.com/AlecAivazis/survey/v2 v2.2.9
 	github.com/andygrunwald/go-jira v1.13.0
+	github.com/coreos/go-semver v0.3.0 // indirect
```
This is a direct dependency, `// indirect` should be removed.

ldelossa replied:
yup you're right. a `go mod tidy` fixed this up.

khos2ow commented on tag_reader.go:
```diff
@@ -71,12 +74,37 @@ func (r *tagReader) ReadAll() ([]*Tag, error) {
+	switch r.sortBy {
+	case "date":
+		r.sortTags(tags)
+		r.assignPreviousAndNextTag(tags)
```
Maybe move `r.assignPreviousAndNextTag(tags)` outside of switch, because it is a shared operation for all.

ldelossa replied:
good one.

khos2ow reviewed (commented):
Apology for late response, yesterday was a bit hectic to say the least! Couple of small comments below, otherwise lgtm.

khos2ow commented:
I finally had a chance to fully compare #78 and this, and except two items, they are pretty much identical, which #78 

1. uses `github.com/hashicorp/go-version` instead of `github.com/coreos/go-semver`
1. defines a CLI flag of `--sort-by-version`

I'm a bit in favor of using sevmer library, but at the same time the other one was approved and supposedly ready to go but now is stalled and in conflict state. I think we can proceed with this PR here as is (after my earlier review comments,) then follow up on #78 to see if the author is willing to rebase, otherwise we can address everything sort related (effectively resolving #112) in another PR.

ldelossa commented:
@khos2ow pr comments addressed in latest push.

clok commented:
Awesome. Great work on this. Appreciate the discourse around the feature. 

@ldelossa this will be in the next release. 

@khos2ow I like your suggested approach. Let's follow up on #78 and #112

ldelossa commented:
My pleasure, thanks for the great tool.

On Sat, Mar 20, 2021 at 6:22 PM Derek Smith ***@***.***>
wrote:

> Merged #124 <https://github.com/git-chglog/git-chglog/pull/124> into
> master.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/git-chglog/git-chglog/pull/124#event-4486069947>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABLBVFSHPWF25FCFWEWCV53TEUNYTANCNFSM4ZLYGIGQ>
> .
>

clok commented:
@ldelossa FYI - [Release v0.12.0](https://github.com/git-chglog/git-chglog/releases/tag/v0.12.0) has rolled out with your feature.

ldelossa commented:
Thanks a ton.
